x/gamm/pool-models/internal/cfmm_common/lp.go

```diff
@@ -0,0 +1,163 @@
+
+	for _, asset := range poolLiquidity {
+		// round down here, due to not wanting to over-exit
+		exitAmt := shareOutRatio.MulInt(asset.Amount).TruncateInt()
+		if exitAmt.LTE(sdk.ZeroInt()) {
+			continue
+		}
+		if exitAmt.GTE(asset.Amount) {
+			return sdk.Coins{}, errors.New("too many shares out")
+		}
+		exitedCoins = exitedCoins.Add(sdk.NewCoin(asset.Denom, exitAmt))
+	}
+
+	return exitedCoins, nil
+}
+
+// MaximalExactRatioJoin LP's the maximal amount of tokens in possible, and returns the number of shares that'd be
+// and how many coins would be left over.
+func MaximalExactRatioJoin(p types.PoolI, ctx sdk.Context, tokensIn sdk.Coins) (numShares sdk.Int, remCoins sdk.Coins, err error) {
+	coinShareRatios := make([]sdk.Dec, len(tokensIn))
+	minShareRatio := sdk.MaxSortableDec
+	maxShareRatio := sdk.ZeroDec()
+
+	poolLiquidity := p.GetTotalPoolLiquidity(ctx)
+	totalShares := p.GetTotalShares()
+
+	for i, coin := range tokensIn {
+		shareRatio := coin.Amount.ToDec().QuoInt(poolLiquidity.AmountOfNoDenomValidation(coin.Denom))
+		if shareRatio.LT(minShareRatio) {
+			minShareRatio = shareRatio
+		}
+		if shareRatio.GT(maxShareRatio) {
+			maxShareRatio = shareRatio
+		}
+		coinShareRatios[i] = shareRatio
+	}
+
+	if minShareRatio.Equal(sdk.MaxSortableDec) {
+		return numShares, remCoins, errors.New("unexpected error in MaximalExactRatioJoin")
+	}
+
+	remCoins = sdk.Coins{}
+	numShares = minShareRatio.MulInt(totalShares).TruncateInt()
+
+	// if we have multiple share values, calculate remainingCoins
+	if !minShareRatio.Equal(maxShareRatio) {
+		// we have to calculate remCoins
+		for i, coin := range tokensIn {
+			// if coinShareRatios[i] == minShareRatio, no remainder
+			if coinShareRatios[i].Equal(minShareRatio) {
+				continue
+			}
+
+			usedAmount := minShareRatio.MulInt(coin.Amount).Ceil().TruncateInt()
+			newAmt := coin.Amount.Sub(usedAmount)
+			// if newAmt is non-zero, add to RemCoins. (It could be zero due to rounding)
+			if !newAmt.IsZero() {
+				remCoins = remCoins.Add(sdk.Coin{Denom: coin.Denom, Amount: newAmt})
+			}
+		}
+	}
+
+	return numShares, remCoins, nil
+}
+
+// We binary search a number of LP shares, s.t. if we exited the pool with the updated liquidity,
+// and swapped all the tokens back to the input denom, we'd get the same amount. (under 0 swap fee)
+// Thanks to CFMM path-independence, we can estimate slippage with these swaps to be sure to get the right numbers here.
+// (by path-independence, swap all of B -> A, and then swap all of C -> A will yield same amount of A, regardless
+// of order and interleaving)
+//
+// This implementation all of pool.GetTotalPoolLiquidity, pool.ExitPool, and pool.SwapExactAmountIn to not
+// require any updates to state, and instead only do updates based upon the pool struct.
+func BinarySearchSingleAssetJoin(
+	pool types.PoolI,
+	tokenIn sdk.Coin,
+	poolWithAddedLiquidityAndShares func(newLiquidity sdk.Coin, newShares sdk.Int) types.PoolI,
```

Comment:
Why do we need to pass in this callback when `pool` that is used by the caller to create a copy is accessible within the scope of this function?

Reply:
We need functions that aren't on the pool API to do this.

Reply:
So that we can customize them for each pool implementation if needed?

Reply:
If that's the reason, a potential alternative is to define a function on the pool interface and call it inside `BinarySearchSingleAssetJoin`

e.g.
```
func BinarySearchSingleAssetJoin(pool types.PoolI, ...) {
    ...
    poolWithUpdatedLiquidity := pool.GetCopyWithLiquidityAndShares(tokenIn, sharesIn)
    ...
}
```

Reply:
Yes this is an alternative. I don't think it is preferrable

Reply:
I think this adding such a method is premature abstraction.